systemd/netplan-sriov-setup.service

```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=Netplan one-shot initial SR-IOV configuration service
+DefaultDependencies=no
+After=systemd-udevd-settle.service
+Wants=systemd-udevd-settle.service
```

Comment:
`systemd-udevd-settle.service` is a deprecated service which should not be used anymore.

Do you mean ordering after `systemd-udev-trigger.service` unit instead?

Overall the unit feels very wrong. For example, shouldn't you simply install a udev rule that is sensitive to network device with sriov functionality, which should then declare `SYSTEMD_WANTS=netplan-sriov-config@%i.service` which would then start that unit and ask netplan to configure sriov on the device? Possibly as a remainafterexit & oneshot service.

Because this needs to happen whenever netplan config appears/ changed, and/or applied whenever network devices appear. Hence, it is more natural to have it udev driven with a custom udev rule.

Or like a networkd-dispatcher hook sensitive to devices appearing? But then not sure if this will work with network-manager.

Reply:
Ah, indeed, thanks for pointing this out!

We basically only need this to run once on boot time in theory, since for a running system it is assumed that, whenever the network/device setup changes, one has to call `netplan apply` explicitly. That being said, I guess this is also an option - and I have an idea how to make an udev rule only for SR-IOV devices. Let me get back to you in a bit.

Reply:
Ok, so even though I found a nice way to write an udev rule that would work for any supported sriov_device (by matching by `ATTRS{sriov_totalvfs}`), I do not think this is a very easy way to go right now.

The problem I see with this is that right now it's most optimal to deal with SR-IOV devices as a whole when they are all available and ready to be configured. I could refactor the SR-IOV initialization so that it could be called for a single device, sure, but then netplan needs to do a lot of parsing and matching multiple times because of how the whole thing has been designed. Like, each time netplan would have to match which VF comes from which PF, then evaluate which PF has been added, then apply the config for the given set of PFs and VFs. Then another device pops up and the same parsing would have to happen for no reason. I prefer to just wait for all the devices to come up and then call SR-IOV initialization once, if possible. Sure, we don't have the fine granularity, but how often do we hot-plug SR-IOV devices?

Another thing I'd have to wrap my head around is the timing. Like, I explicitly set `systemd-udevd-settle` as the target as I need the SR-IOV initialization to happen *after* all the interface renaming has happened. I don't know these parts too well, but how could I make sure that all the related network device renaming per netplan's set-name have been done and only then do my SR-IOV initialization? Since this is the goal.

Reply:
On the other hand, guess I'd live with the extra parsing, but I wonder if there's a way to just make sure the interface has been renamed when the netplan trigger is happening. Let me experiment some more.

Reply:
@sil2100 you can make netplan generate create a target. Then from neptlan yaml, generate dependencies on all of the sriov devices that netplan says to configure. Then make the target depend on all of those devices. Then make the unit that does `netplan-sriov-setup` to be after the target, this way it will be run once, after all the sriov devices are up.

Reply:
Argh, this is terrible. The worst thing is that even though I can get the right timing done for networkd (as I can create udev rules that would fire after the interfaces have final names), I can't really do that easily for NetworkManager. And it's important in this approach as the VFs need to exist when NetworkManager starts configuring them as otherwise they won't get all the usual settings. So argh, I'd probably have to restart NM after applying SR-IOV config, which really sucks.

Ok, so maybe instead I should revert to my original idea of running SR-IOV initialization really early, before the renames. I didn't like that approach as even though it would look really nice during boot, on runtime its a bit confusing as some interfaces can have new names and others - old. But maybe it makes sense to make this smarter? Like, take into consideration process_link_changes() and then check in the SR-IOV code which name should be considered? Let me look into that then.